File: server/python/src/synckit_server/auth/jwt.py
"""
JWT Authentication Module

Matches TypeScript reference: server/typescript/src/auth/jwt.ts
"""

from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Optional

from jose import jwt, ExpiredSignatureError, JWTError

from ..config import settings


@dataclass
class DocumentPermissions:
    """Document-level permissions"""

    can_read: list[str] = field(default_factory=list)  # Document IDs user can read
    can_write: list[str] = field(default_factory=list)  # Document IDs user can write
    is_admin: bool = False  # Admin has access to all documents


@dataclass
class TokenPayload:
    """JWT Token Payload"""

    user_id: str
    permissions: DocumentPermissions
    email: Optional[str] = None
    iat: Optional[int] = None  # Issued at
    exp: Optional[int] = None  # Expiration


def verify_token(token: str) -> Optional[TokenPayload]:
    """
    Verify and decode JWT token.

    Must match TypeScript behavior exactly:
    - Validate signature with secret
    - Check expiration
    - Return None on any error (no exceptions propagated)

    Args:
        token: JWT token string

    Returns:
        TokenPayload if valid, None if invalid/expired
    """
    try:
        decoded = jwt.decode(
            token,
            settings.jwt_secret,
            algorithms=[settings.jwt_algorithm],
        )

        # Extract permissions
        perms_data = decoded.get("permissions", {})
        permissions = DocumentPermissions(
            can_read=perms_data.get("canRead", []),
            can_write=perms_data.get("canWrite", []),
            is_admin=perms_data.get("isAdmin", False),
        )

        return TokenPayload(
            user_id=decoded.get("userId", ""),
            email=decoded.get("email"),
            permissions=permissions,
            iat=decoded.get("iat"),
            exp=decoded.get("exp"),
        )
    except ExpiredSignatureError:
        # Token expired - matches TypeScript jwt.TokenExpiredError
        return None
    except JWTError:
        # Invalid token - matches TypeScript jwt.JsonWebTokenError
        return None
    except Exception:
        # Any other error
        return None


def generate_access_token(
    user_id: str,
    permissions: DocumentPermissions,
    email: Optional[str] = None,
    expires_in_hours: Optional[int] = None,
) -> str:
    """
    Generate JWT access token.

    Args:
        user_id: User identifier
        permissions: Document permissions
        email: Optional email address
        expires_in_hours: Token expiration in hours (default from config)

    Returns:
        JWT token string
    """
    if expires_in_hours is None:
        expires_in_hours = settings.jwt_expiration_hours

    now = datetime.now(timezone.utc)
    expire = now + timedelta(hours=expires_in_hours)

    payload = {
        "userId": user_id,
        "permissions": {
            "canRead": permissions.can_read,
            "canWrite": permissions.can_write,
            "isAdmin": permissions.is_admin,
        },
        "iat": int(now.timestamp()),
        "exp": int(expire.timestamp()),
    }

    if email:
        payload["email"] = email

    return jwt.encode(payload, settings.jwt_secret, algorithm=settings.jwt_algorithm)


def generate_refresh_token(user_id: str, expires_in_days: int = 7) -> str:
    """
    Generate JWT refresh token.

    Args:
        user_id: User identifier
        expires_in_days: Token expiration in days (default 7)

    Returns:
        JWT token string
    """
    now = datetime.now(timezone.utc)
    expire = now + timedelta(days=expires_in_days)

    payload = {
        "userId": user_id,
        "iat": int(now.timestamp()),
        "exp": int(expire.timestamp()),
    }

    return jwt.encode(payload, settings.jwt_secret, algorithm=settings.jwt_algorithm)


def generate_tokens(
    user_id: str,
    email: str,
    permissions: DocumentPermissions,
) -> dict[str, str]:
    """
    Generate both access and refresh tokens for a user.

    Args:
        user_id: User identifier
        email: User email
        permissions: Document permissions

    Returns:
        Dict with 'accessToken' and 'refreshToken'
    """
    return {
        "accessToken": generate_access_token(user_id, permissions, email),
        "refreshToken": generate_refresh_token(user_id),
    }


def decode_token(token: str) -> Optional[TokenPayload]:
    """
    Decode token without verification (for debugging).

    Args:
        token: JWT token string

    Returns:
        TokenPayload if decodable, None otherwise
    """
    try:
        # Decode without verification
        decoded = jwt.decode(
            token,
            settings.jwt_secret,
            algorithms=[settings.jwt_algorithm],
            options={"verify_signature": False, "verify_exp": False},
        )

        perms_data = decoded.get("permissions", {})
        permissions = DocumentPermissions(
            can_read=perms_data.get("canRead", []),
            can_write=perms_data.get("canWrite", []),
            is_admin=perms_data.get("isAdmin", False),
        )

        return TokenPayload(
            user_id=decoded.get("userId", ""),
            email=decoded.get("email"),
            permissions=permissions,
            iat=decoded.get("iat"),
            exp=decoded.get("exp"),
        )
    except Exception:
        return None

File: server/go/internal/auth/jwt.go
// Package auth provides JWT authentication for SyncKit server.
// Matches TypeScript reference: server/typescript/src/auth/jwt.ts
package auth

import (
	"errors"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// DocumentPermissions represents document-level permissions
type DocumentPermissions struct {
	CanRead  []string `json:"canRead"`  // Document IDs user can read
	CanWrite []string `json:"canWrite"` // Document IDs user can write
	IsAdmin  bool     `json:"isAdmin"`  // Admin has access to all documents
}

// TokenPayload represents JWT token claims
type TokenPayload struct {
	UserID      string              `json:"userId"`
	Email       string              `json:"email,omitempty"`
	Permissions DocumentPermissions `json:"permissions"`
	jwt.RegisteredClaims
}

// Errors for JWT validation
var (
	ErrInvalidToken = errors.New("invalid token")
	ErrExpiredToken = errors.New("token expired")
	ErrShortSecret  = errors.New("JWT secret must be at least 32 characters")
)

// VerifyToken verifies and decodes a JWT token.
//
// Must match TypeScript behavior exactly:
// - Validate signature with secret
// - Check expiration
// - Return nil on any error (no exceptions propagated)
func VerifyToken(tokenString, secret string) (*TokenPayload, error) {
	// Validate secret length (security requirement)
	if len(secret) < 32 {
		return nil, ErrShortSecret
	}

	token, err := jwt.ParseWithClaims(tokenString, &TokenPayload{}, func(token *jwt.Token) (interface{}, error) {
		// Validate signing method
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(secret), nil
	})

	if err != nil {
		if errors.Is(err, jwt.ErrTokenExpired) {
			return nil, ErrExpiredToken
		}
		return nil, ErrInvalidToken
	}

	if claims, ok := token.Claims.(*TokenPayload); ok && token.Valid {
		return claims, nil
	}

	return nil, ErrInvalidToken
}

// GenerateAccessToken generates a JWT access token.
func GenerateAccessToken(userID string, email string, permissions DocumentPermissions, secret string, expiresIn time.Duration) (string, error) {
	if len(secret) < 32 {
		return "", ErrShortSecret
	}

	now := time.Now()
	claims := &TokenPayload{
		UserID:      userID,
		Email:       email,
		Permissions: permissions,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(now.Add(expiresIn)),
			IssuedAt:  jwt.NewNumericDate(now),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(secret))
}

// GenerateRefreshToken generates a JWT refresh token.
func GenerateRefreshToken(userID, secret string, expiresIn time.Duration) (string, error) {
	if len(secret) < 32 {
		return "", ErrShortSecret
	}

	now := time.Now()
	claims := jwt.RegisteredClaims{
		Subject:   userID,
		ExpiresAt: jwt.NewNumericDate(now.Add(expiresIn)),
		IssuedAt:  jwt.NewNumericDate(now),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(secret))
}

// GenerateTokens generates both access and refresh tokens for a user.
func GenerateTokens(userID, email string, permissions DocumentPermissions, secret string) (accessToken, refreshToken string, err error) {
	accessToken, err = GenerateAccessToken(userID, email, permissions, secret, 24*time.Hour)
	if err != nil {
		return "", "", err
	}

	refreshToken, err = GenerateRefreshToken(userID, secret, 7*24*time.Hour)
	if err != nil {
		return "", "", err
	}

	return accessToken, refreshToken, nil
}

// DecodeTokenWithoutVerification decodes token without verification (for debugging).
func DecodeTokenWithoutVerification(tokenString string) (*TokenPayload, error) {
	token, _, err := new(jwt.Parser).ParseUnverified(tokenString, &TokenPayload{})
	if err != nil {
		return nil, err
	}

	if claims, ok := token.Claims.(*TokenPayload); ok {
		return claims, nil
	}

	return nil, ErrInvalidToken
}
